sdk/python/kfp/v2/components/experimental/component_spec.py

```diff
@@ -0,0 +1,332 @@
+    self.metdata = metadata
+
+  def _validate_input_output_uniqueness(
+      self,
+      input_or_output_specs: Optional[Union[Sequence[InputSpec],
+                                            Sequence[OutputSpec]]],
+  ) -> Optional[Union[Sequence[InputSpec], Sequence[OutputSpec]]]:
+    names = set()
+    for spec in input_or_output_specs or []:
+      if spec.name in names:
+        input_or_output = 'input' if isinstance(spec, InputSpec) else 'output'
+        raise ValueError(f'Non-unique {input_or_output} name "{spec.name}".')
+      names.add(spec.name)
+    return input_or_output_specs
+
+  def _validate_placeholders(
+      self,
+      implementation: Union[ContainerSpec, ImporterSpec, DagSpec],
+  ) -> Union[ContainerSpec, DagSpec]:
+    if not isinstance(implementation, ContainerSpec):
+      return implementation
+
+    input_names = [input_spec.name for input_spec in self.input_specs or []]
+    output_names = [output_spec.name for output_spec in self.output_specs or []]
+
+    for arg in [
+        *(implementation.commands or []), *(implementation.arguments or [])
+    ]:
+      if isinstance(
+          arg,
+          (InputValuePlaceholder, InputPathPlaceholder, InputUriPlaceholder)):
+        if arg.input_or_output_name not in input_names:
+          raise ValueError(f'Argument "{arg}" references non-existing input.')
+      elif isinstance(arg, (OutputPathPlaceholder, OutputUriPlaceholder)):
+        if arg.input_or_output_name not in output_names:
+          raise ValueError(f'Argument "{arg}" references non-existing output.')
+      # TODO(chensun): revisit if we need to support IfPlaceholder,
+      # ConcatPlaceholder, etc. in the new format.
+      elif not isinstance(arg, str):
+        raise ValueError(f'Unpexpected argument "{arg}".')
+    return implementation
+
+  @classmethod
+  def from_v1_component_spec(
+      cls, v1_component_spec: v1_components.ComponentSpec) -> 'ComponentSpec':
+    raise NotImplementedError
+
+  def to_v1_component_spec(self) -> v1_components.ComponentSpec:
+    if isinstance(self.implementation, DagSpec):
+      raise NotImplementedError
+
+    def _transform_arg(arg: Union[str, BasePlaceholder]) -> Any:
+      if isinstance(arg, str):
+        return arg
+      elif isinstance(arg, InputValuePlaceholder):
+        return v1_components.InputValuePlaceholder(arg.input_or_output_name)
+      elif isinstance(arg, InputPathPlaceholder):
+        return v1_components.InputPathPlaceholder(arg.input_or_output_name)
+      elif isinstance(arg, InputUriPlaceholder):
+        return v1_components.InputUriPlaceholder(arg.input_or_output_name)
+      elif isinstance(arg, OutputPathPlaceholder):
+        return v1_components.OutputPathPlaceholder(arg.input_or_output_name)
+      elif isinstance(arg, OutputUriPlaceholder):
+        return v1_components.OutputUriPlaceholder(arg.input_or_output_name)
+      else:
+        raise ValueError(
```

Comment:
Shouldn't we also check for If and Concat placeholders here?

Reply:
We should still be able to load them though. A TODO is fine. Maybe add a tracking bug so we don't forget as well.

Reply:
Opened a bug.